[Enhancement]enable fe to list all file in hdfs recursively

adzfolc opened the pull request:
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #10238

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [x] I have added user document for my new feature or new function
If user set `recursive_dir_search_enabled = true` in fe.conf, fe will scan hdfs files recursively; otherwise fe won't change old scan policy.

adzfolc committed:
1. enable fe to list all file in hdfs recursively

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 1 / 1 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|1|1|100.00%|[]|

DorianZheng commented on fe/fe-core/src/main/java/com/starrocks/external/hive/HiveMetaClient.java:
```diff
@@ -682,7 +682,8 @@ public List<HdfsFileDesc> getHdfsFileDescs(String dirPath, boolean isSplittable,
         try {
-            RemoteIterator<LocatedFileStatus> blockIterator = fileSystem.listLocatedStatus(new Path(uri.getPath()));
+            // files in hdfs may have multiple directories, so we need to list all files in hdfs recursively here
+            RemoteIterator<LocatedFileStatus> blockIterator = fileSystem.listFiles(new Path(uri.getPath()), true);
```
this is not a bug, it's a enhancement. this should be configurable.

adzfolc replied:
> this is not a bug, it's a enhancement. this should be configurable.

After consideration,it's more appropriate to put the para into `fe.conf`.

adzfolc committed:
1. enable fe to list all file in hdfs recursively

adzfolc committed:
1. change to session variable

adzfolc commented:
Reminder:need to backport to branch-2.3

adzfolc committed:
Merge branch 'main' into fix-enable-fe-scan-hdfs-file-recursively

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 8 (50.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|1|4|25.00%|[1101, 1105, 1106]|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|3|4|75.00%|[680]|

adzfolc committed:
1. change config from session variable to fe config

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 4 / 5 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|3|4|75.00%|[679]|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|[]|

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/common/Config.java:
```diff
@@ -1698,4 +1698,7 @@ public class Config extends ConfigBase {
     public static boolean enable_new_publish_mechanism = false;
+
+    @ConfField(mutable = true)
+    public static boolean recursive_dir_search_enabled = true;
```
default value is false

adzfolc replied:
> default value is false

changed

adzfolc committed:
1. change recursive_dir_search_enabled default value to false

wangruin commented:
run starrocks_clang-tidy

adzfolc commented:
run starrocks_fe_unittest

adzfolc commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 4 / 5 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|3|4|75.00%|[681]|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|[]|

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/external/hive/HiveMetaClient.java:
```diff
@@ -673,7 +673,13 @@ public List<HdfsFileDesc> getHdfsFileDescs(String dirPath, boolean isSplittable,
+            }
             while (blockIterator.hasNext()) {
                 LocatedFileStatus locatedFileStatus = blockIterator.next();
                 if (!isValidDataFile(locatedFileStatus)) {
```
isValidDataFile() will skip the dir, you could check it.

adzfolc replied:
> isValidDataFile() will skip the dir, you could check it.

After check, now hdfs directory will be scanned recursively by call getHdfsFileDescs.

adzfolc committed:
1. enable fe to recursively scan hdfs directories
2. if user set `recursive_dir_search_enabled = true` in fe.conf, fe will scan hdfs files recursively

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 8 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|5|7|71.43%|[681, 688]|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|[]|

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/external/hive/HiveMetaClient.java:
```diff
@@ -673,11 +673,19 @@ public List<HdfsFileDesc> getHdfsFileDescs(String dirPath, boolean isSplittable,
                 LocatedFileStatus locatedFileStatus = blockIterator.next();
                 if (!isValidDataFile(locatedFileStatus)) {
                     continue;
+                } else if (locatedFileStatus.isDirectory() && Config.recursive_dir_search_enabled) {
```
I have checked the listFiles(true)'s behavior. It will return the files and the subDir's files. So I think there is no need to check the isDir?

adzfolc replied:
> I have checked the listFiles(true)'s behavior. It will return the files and the subDir's files. So I think there is no need to check the isDir?

As this pr's target, fe should scan hdfs files recursively only when config set `recursive_dir_search_enabled` to true.So,if locatedFileStatus is a dir, and fe enable `recursive_dir_search_enabled`, we check dir here. If recursive_dir_search_enabled is false, fe shouldn't scan sub dir of hdfs.

adzfolc commented:
@DorianZheng @caneGuy PTAL

wanpengfei-git commented:
run starrocks_admit_test

adzfolc commented:
@Mergifyio backport branch-2.3

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 8 (75.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/hive/HiveMetaClient.java|5|7|71.43%|[681, 688]|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|[]|

Smith-Cruise commented:
@Mergifyio backport branch-2.4
